dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorIMM.java

```diff
@@ -24,98 +25,102 @@
 
 class VisitorIMM extends VisitorBase {
 
-	protected VisitorIMM(boolean forceStart) {
-		super(forceStart);
-	}
+    protected VisitorIMM(boolean forceStart) {
+        super(forceStart);
+    }
 
-	@Override
-	public List<Event> visitLoad(Load e) {
-		String mo = e.getMo();
+    @Override
+    public List<Event> visitLoad(Load e) {
+        // FIXME: It is weird to compile a core-level load by transforming its tagging.
+        final MemoryOrder mo = e.getMetadata(MemoryOrder.class);
+        final boolean isNonAtomic = (mo == null || mo.value().equals(C11.NONATOMIC));
         return eventSequence(
-				newLoad(e.getResultRegister(), e.getAddress(), mo.isEmpty() || mo.equals(C11.NONATOMIC) ? C11.MO_RELAXED : mo)
+                newLoadWithMo(e.getResultRegister(), e.getAddress(), isNonAtomic ? C11.MO_RELAXED : mo.value())
         );
-	}
+    }
 
-	@Override
-	public List<Event> visitStore(Store e) {
-		String mo = e.getMo();
+    @Override
+    public List<Event> visitStore(Store e) {
+        // FIXME: It is weird to compile a core-level load by transforming its tagging.
+        final MemoryOrder mo = e.getMetadata(MemoryOrder.class);
+        final boolean isNonAtomic = (mo == null || mo.value().equals(C11.NONATOMIC));
         return eventSequence(
-				newStore(e.getAddress(), e.getMemValue(), mo.isEmpty() || mo.equals(C11.NONATOMIC) ? C11.MO_RELAXED : mo)
+                newStoreWithMo(e.getAddress(), e.getMemValue(), isNonAtomic ? C11.MO_RELAXED : mo.value())
         );
-	}
+    }
 
-	@Override
-	public List<Event> visitCreate(Create e) {
-        Store store = newStore(e.getAddress(), e.getMemValue(), C11.MO_RELEASE);
+    @Override
+    public List<Event> visitCreate(Create e) {
+        Store store = newStoreWithMo(e.getAddress(), e.getMemValue(), C11.MO_RELEASE);
         store.addTags(C11.PTHREAD);
-        
+
         return eventSequence(
                 store
         );
-	}
+    }
 
-	@Override
-	public List<Event> visitEnd(End e) {
+    @Override
+    public List<Event> visitEnd(End e) {
         //TODO boolean
```

Comment:
what is this TODO?

Reply:
This is a TODO from the #457 PR. 
I think it says that we eventually want to use boolean variables to communicate thread creation (since we only use values `0` and `1` anyway).